docs/utxo-hd/v0.1/for-developers.md

```diff
@@ -0,0 +1,456 @@
+`ImmutableTip` and `L1`, `D2` is the delta between `L1` and `L2`, and so on.
+Notice the existence of deltas with negative numbers. These are the deltas
+belonging to the (already garbage-collected as they are not needed) immutable
+states between the `LastFlushedState` and the `ImmutableTip`.
+
+The sequence of differences is kept in a `FingerTree` as mentioned above.
+Pushing and splitting this sequence is one of the main causes of speed loss in
+the node. In order to not traverse the whole `FingerTree` every time we need to
+use the deltas, the modified `FingerTree` keeps a root measure that is the
+combination of all the deltas so that we can just apply that, but it comes at a
+cost as we are pushing only one element into the `FingerTree` after each block,
+and splitting a `FingerTree` is already a costly operation.
+
+One might think: why then keep this sequence at all? There are the reasons
+mentioned at the beginning of this section but there is also a crucial one for
+the well-functioning of the node: we need to be able to quickly consult any of
+the last `k` ledger states in order to evaluate forks and roll-back the chain if
+necessary. Using this sequence, we are able to re-construct a `k`-recent Ledger
+state suitable for these purpose.
+
+### The Backing Store (on-disk store)
+
+As you might have noticed in the diagram above, we only keep track of
+differences in the `DbChangelog`, but at some point these differences must have
+a base value/anchor on top of which we can apply them. This anchor is an
+instance of the whole UTxO set (in particular the UTxO set at the last flushed
+state) that we keep in what we named the [_backing
+store_](../../../ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/BackingStore.hs).
+In order to manipulate the Backing Store, we produce a
+`BackingStoreValueHandle`, see the previous link for more details.
+
+On top of the values that exist in the backing store, the sequence of
+differences in the LedgerDB can be applied to create a UTxO set at a specific
+slot younger than the one in the backing store.
+
+Functionally it is just a key-value store (plus the `SlotNo` of the last
+flushed state), and we provide two implementations:
+
+- [InMemory](../../../ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/InMemory.hs):
+  is a `Data.Map` that lives in a `TVar`.
+- [LMDB](../../../ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs)
+  uses [LMDB](http://www.lmdb.tech/doc/) bindings to keep a database in a file in the disk.
+
+To be precise, only the LMDB implementation will decrease the memory usage by
+moving this instance of the UTxO set to disk, but it comes at a cost as I/O has
+to be performed to read and write values. We expect SPOs with powerful machines
+to use the InMemory implementation and every other user to move to the LMDB
+implementation.
```

Comment:
Maybe mention that a bespoke back-end is also being implemented but won't be part of the first release?

And/or explain why LMDB is sufficient now but might not remain sufficient in the (near) future?

Reply:
Added